### ISSUE TITLE
Sources filtered by criterion

### ISSUE THREAD

**Issue by sahalali** [culturecreates/footlight-aggregator#143]

**Comment by troughc** [comment]
https://placedesarts.com/fr/evenement/half-moon-run-osm-pop

**Comment by troughc** [comment]
https://kg.artsdata.ca/entity?uri=http%3A%2F%2Fkg.artsdata.ca%2Fresource%2FK23-1785

**Comment by dev-aravind** [comment]
Assigning to @saumier to provide a sample source for @dev-aravind to test the filters.

**Comment by saumier** [comment]
@dev The sample [event](https://kg.artsdata.ca/entity?uri=http%3A%2F%2Fkg.footlight.io%2Fresource%2Fplacedesarts-com_half-moon-run-osm-pop) is now linked to the artists Adam Johnson 
[ad:K5-513](https://kg.artsdata.ca/entity?uri=http%3A%2F%2Fkg.artsdata.ca%2Fresource%2FK5-513).  Please try to load the source placedesarts-com from Artsdata.

**Comment by dev-aravind** [comment]
@saumier 

Proposed aggregator algorithm to filter entities in Artsdata:

- Fetch all the entities in the source for which we have a filter for (if we are filtering events by some places we will fetch all the places first)

- Filter the entities by filters mentioned in the mapping file, for example if the filter is :

  {
    "entityType" : "PLACE",
    "artsdataFilters":[
      {
        "inputProperty":["address.addressRegion.@none"],
        "includePatterns" : ["NB"],
        "excludePatterns" : [] 
      }
    ]
  }

we will get places from the region "NB" only

- Fetch events and check if the location in each individual event is present in any of the filtered places.

- If yes, the event gets imported, else it gets skipped. 


I would have to write 2 SPARQLs to make this work, one to get all the places in case of a place filter and one to get all the people/organization in case of a organizer/performer/sponsor filter. Please review this and let me know what you think.

**Comment by saumier** [comment]
@dev-aravind Thanks for writing down your psuedocode. It is very clear.  

I have a few concerns.  Currently the SPARQL to fetch events from Artsdata is the same for all sources.  It is tempting to customize this SPARQL. For example, the SPARQL could fetch only the events`?event  s:location/s:address/s:addressRegion ?region . filter(str(?region) = "NB")`.  When I discussed this direction with Suhail last year we decided to keep one SPARQL for all sources so that we could maintain the SPARQL without having to manage many duplicate SPARQLs with small differences per calendar.  So I like your idea of 2 additional SPARQLs for places and agents(person or organization).  But is there a more elegant solution (less code)? There are several ideas that come to mind but I haven't thought through them. One idea is to load the data into a local graph in Aggregator (using Javascript RDF) and then apply the filter in a single step. Another idea is divide the calls to Artsdata into 2 steps: get a list of events IDs, then load the data for each event. If we use those 2 steps, the first could be filtered to only list the events to import (i.e. events in NB), and the second could always load the event data in the same way to be easy to maintain.  The second step takes as input the list of events from the first step. Do you think that would work?

Something else I'd like you to know about. The current SPARQL to load events from Artsdata will change sometime in the next 6 months. The plan is to simplify it by using the minted events in Artsdata. We are going through a transition at the moment by using the SPARQL to merge on-the-fly events to use minted IDs.  

To summarize, I am proposing the following pattern: 
1. SPARQL to get the list of events (same for all calendars)
2. SPARQL to filter the list of events (i.e. custom filter(str(?region) = "NB"). It could be generated from the calendar's filter configuration.
3. SPARQL to get event data for each event in the list (same for all calendars).

**Comment by dev-aravind** [comment]
@saumier The [event](https://staging.cms.footlight.io/dashboard/63457978637093005071a053/events/676164b5f1426900a2e47e9c) is now imported into the staging CMS.

**Comment by saumier** [comment]
@sahalali Lets review.

**Comment by saumier** [comment]
@sahalali I think this feature enhancement is best achieved using the enhancements we are making to the reconciliation server (match, suggest, data extension). https://github.com/culturecreates/artsdata-reconciliation/issues/26 

The idea is to define aggregators using the reconciliation service. This would eventually replace all aggregators.  CMS would need a new UI screen to configure aggregators.

In this user story, the Reconciliation server would be called with a request to match all People with residency in Laval (property discovered using the suggest service) and then call the data extension API to batch load the data from Artsdata.  

What do you think?

**Comment by sahalali** [comment]
We need to wait until the recon features (match, suggest, data extension) are complete. The UI screens need to be designed for this.

**Comment by troughc** [comment]
@sahalali @saumier I recommend that any UI development would be done in a separate ticket. For this issue, let's aim to have an aggregator import these types of events (without being managed by the Admin)
